pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_reconcile.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apireconciler
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/apidefinition"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+func (c *APIReconciler) reconcile(ctx context.Context, apiExport *apisv1alpha1.APIExport, apiDomainKey dynamiccontext.APIDomainKey) error {
+	if apiExport == nil || apiExport.Status.IdentityHash == "" {
+		c.mutex.RLock()
+		_, found := c.apiSets[apiDomainKey]
+		c.mutex.RUnlock()
+
+		if !found {
+			klog.V(3).Infof("No APIs found for API domain key %s", apiDomainKey)
+			return nil
+		}
+
+		// remove the APIDomain
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+		klog.V(2).Infof("Deleting APIs for API domain key %s", apiDomainKey)
+		delete(c.apiSets, apiDomainKey)
+		return nil
+	}
+
+	c.mutex.RLock()
+	oldSet := c.apiSets[apiDomainKey]
+	c.mutex.RUnlock()
+
+	// collect APIResourceSchemas
+	schemaIdentity := map[string]string{}
+	apiResourceSchemas := make([]*apisv1alpha1.APIResourceSchema, 0, len(apiExport.Spec.LatestResourceSchemas))
```

Comment:
Not to be a style bandit here but why bother with this over `var apiResourceSchemas []*apisv1alpha1.APIResourceSchema` ?

Reply:
If it's not found, this is the wrong size ...

Reply:
slightly over-estimated. Wrong is a strong word.